### ISSUE TITLE
Traefik doesn't give certificate to the API endpoint

### ISSUE THREAD

**Issue by sunsh1newarrior** [louis-lau/wildduck-dockerized#19]
After initial configuration, traefik generates a SSL certificate for 80 HTTP port, but not for 8080 API port.
I can't use the API because of that, REST client complains about self-signed certificate.

**Comment by louis-lau** [comment]
By default the api is exposed without authentication on localhost over HTTP. Getting a self signed cert from that would be quite hard :), so I'm not completely sure what you're seeing.

If you do want to expose the api to the world, you can set up api authentication in the wildduck config, and add some traefik labels to make traefik proxy it.

Make sure you have a good reason to expose the api to the world though, in most setups it's unnecessary. (That's why it's not exposed by default)

Also don't expose it without authentication, that would give everyone access to all your email.

**Comment by gmp2k** [comment]
>If you do want to expose the api to the world, you can set up api authentication in the wildduck config

I've already done that. I'm just curious how to make API connection secure "in a correct way".
Furthermore, I tried to mess with traefik labels in docker-compose file, no luck. Of course, I can just extract existing certificates from traefik's acme.json in a docker volume and point to them in api.toml config file, but updating them regularly is a pain and this solution doesn't seem right.

**Comment by louis-lau** [comment]
Can you send me what you tried with the labels? Maybe I can tell you where you went wrong.

**Comment by gmp2k** [comment]
Okay, here it is

```yml
version: "3.8"

volumes:
  mongo:
  redis:
  traefik:

services: 
  wildduck:
    image: nodemailer/wildduck
    restart: unless-stopped
    ports:
      - "8080:8080"
      - "143:143"
      - "110:110"
    depends_on:
      - mongo
      - redis
    volumes:
      - ./config/wildduck:/wildduck/config
    labels:
      traefik.enable: true
      traefik.tcp.routers.wildduck-imaps.entrypoints: imaps
      traefik.tcp.routers.wildduck-imaps.rule: HostSNI(`${HOSTNAMES}`)
      traefik.tcp.routers.wildduck-imaps.tls.certresolver: letsencrypt
      traefik.tcp.routers.wildduck-imaps.service: wildduck-imaps
      traefik.tcp.services.wildduck-imaps.loadbalancer.server.port: 143
      traefik.tcp.routers.wildduck-pop3s.entrypoints: pop3s
      traefik.tcp.routers.wildduck-pop3s.rule: HostSNI(`${HOSTNAMES}`)
      traefik.tcp.routers.wildduck-pop3s.tls.certresolver: letsencrypt
      traefik.tcp.routers.wildduck-pop3s.service: wildduck-pop3s
      traefik.tcp.services.wildduck-pop3s.loadbalancer.server.port: 110
      traefik.tcp.routers.wildduck-apis.entrypoints: apis
      traefik.tcp.routers.wildduck-apis.rule: HostSNI(`${HOSTNAMES}`)
      traefik.tcp.routers.wildduck-apis.tls.certresolver: letsencrypt
      traefik.tcp.routers.wildduck-apis.service: wildduck-apis
      traefik.tcp.services.wildduck-apis.loadbalancer.server.port: 8080

  wildduck-webmail:
    image: nodemailer/wildduck-webmail
    command: 
      - ""
    restart: unless-stopped
    depends_on:
      - mongo
      - redis
      - wildduck
    volumes:
      - ./config/wildduck-webmail:/app/config
    labels:
      traefik.enable: true
      traefik.http.routers.wildduck-webmail.entrypoints: https
      traefik.http.routers.wildduck-webmail.rule: Host(`${HOSTNAMES}`)
      traefik.http.routers.wildduck-webmail.tls.certresolver: letsencrypt
      traefik.http.services.wildduck-webmail.loadbalancer.server.port: 3000

  zonemta:
    image: nodemailer/zonemta-wildduck
    restart: unless-stopped
    ports:
      - "587:587"
    depends_on: 
      - mongo
      - redis
    volumes:
      - ./config/zone-mta:/app/config
    labels:
      traefik.enable: true
      traefik.tcp.routers.zonemta.entrypoints: smtps
      traefik.tcp.routers.zonemta.rule: HostSNI(`${HOSTNAMES}`)
      traefik.tcp.routers.zonemta.tls.certresolver: letsencrypt
      traefik.tcp.services.zonemta.loadbalancer.server.port: 587

  haraka:
    image: nodemailer/haraka-wildduck
    restart: unless-stopped
    ports:
      - "25:25"
    depends_on:
      - mongo
      - redis
      - rspamd
    volumes:
      - ./config/haraka:/app/config

  rspamd:
    image: nodemailer/rspamd
    restart: unless-stopped
    depends_on:
      - redis
    volumes:
      - ./config/rspamd/override.d:/etc/rspamd/override.d
      - ./config/rspamd/worker-normal.conf:/etc/rspamd/worker-normal.conf
      - ./config/rspamd/local.d:/etc/rspamd/local.d

  mongo:
    image: mongo
    restart: unless-stopped
    volumes:
      - mongo:/data/db
    
  redis:
    image: redis:alpine
    restart: unless-stopped
    volumes:
      - redis:/data

  traefik:
    image: traefik
    restart: unless-stopped
    command:
      - "--entrypoints.http.address=:80"
      - "--entrypoints.https.address=:443"
      - "--entrypoints.apis.address=:8080"
      - "--entrypoints.imaps.address=:993"
      - "--entrypoints.pop3s.address=:995"
      - "--entrypoints.smtps.address=:465"
      - "--providers.docker"
      - "--providers.docker.exposedByDefault=false"
      - "--certificatesresolvers.letsencrypt.acme.email=${ACME_EMAIL}"
      - "--certificatesresolvers.letsencrypt.acme.storage=/data/acme.json"
      - "--certificatesresolvers.letsencrypt.acme.tlschallenge=true"
    labels:
      traefik.enable: true
      traefik.http.middlewares.redirect-to-https.redirectscheme.scheme: https
      traefik.http.routers.redirs.entrypoints: http
      traefik.http.routers.redirs.middlewares: redirect-to-https
      traefik.http.routers.redirs.rule: hostregexp(`{host:.+}`)
    ports:
      - 80:80/tcp
      - 443:443/tcp
      - 993:993/tcp
      - 995:995/tcp
      - 465:465/tcp
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock:ro
      - traefik:/data
 ```

**Comment by louis-lau** [comment]
Alright, so the entrypoint should be `https`, not `apis`. You should also use `Host` instead of `HostSNI`, and use another hostname. As the webmail is already running on that hostname. If you do need to use the same hostname you can do some complicated path rewriting stuff with traefik I think, but I recommend just using another hostname for either the webmail or the api. I also don't think the `service` label is needed here.

So basically your mistake was copying the labels from the TCP imap and smtp proxy. You should look at the way the webmail is configured, that's already an HTTP proxy :P. Just make sure to not use the same hostnames for them.

**Comment by gmp2k** [comment]
Well, I'll try to rewrite traefik entries. Thank you for the advice.
